src/mlpack/tests/main_tests/image_converter_test.cpp

```diff
@@ -0,0 +1,190 @@
+                         ImageConverterTestFixture);
+
+BOOST_AUTO_TEST_CASE(LoadImageTest)
+{
+  SetInputParam<vector<string>>("input", {"test_image.png", "test_image.png"});
+  SetInputParam("height", 50);
+  SetInputParam("width", 50);
+  SetInputParam("channel", 3);
+
+  mlpackMain();
+  arma::mat output = CLI::GetParam<arma::mat>("output");
+  // width * height * channels.
+  BOOST_REQUIRE_EQUAL(output.n_rows, 50 * 50 * 3);
+  BOOST_REQUIRE_EQUAL(output.n_cols, 2);
+}
+
+BOOST_AUTO_TEST_CASE(SaveImageTest)
+{
+  arma::mat testimage = arma::conv_to<arma::mat>::from(
+      arma::randi<arma::Mat<unsigned char>>((5 * 5 * 3), 2));
+  SetInputParam<vector<string>>("input", {"test_image777.png",
+      "test_image999.png"});
+  SetInputParam("height", 5);
+  SetInputParam("width", 5);
+  SetInputParam("channel", 3);
+  SetInputParam("save", true);
+  SetInputParam("dataset", testimage);
+  mlpackMain();
+
+  CLI::ClearSettings();
+  CLI::RestoreSettings(testName);
+
+  SetInputParam<vector<string>>("input", {"test_image777.png",
+    "test_image999.png"});
+  SetInputParam("height", 5);
+  SetInputParam("width", 5);
+  SetInputParam("channel", 3);
+
+  mlpackMain();
+  arma::mat output = CLI::GetParam<arma::mat>("output");
+  BOOST_REQUIRE_EQUAL(output.n_rows, 5 * 5 * 3);
+  BOOST_REQUIRE_EQUAL(output.n_cols, 2);
+  for (size_t i = 0; i < output.n_elem; ++i)
+    BOOST_REQUIRE_CLOSE(testimage[i], output[i], 1e-5);
+}
+
+/**
+ * Check Saved model is working.
+ */
+BOOST_AUTO_TEST_CASE(SavedModelTest)
+{
+  SetInputParam<vector<string>>("input", {"test_image.png", "test_image.png"});
+  SetInputParam("height", 50);
+  SetInputParam("width", 50);
+  SetInputParam("channel", 3);
+
+  mlpackMain();
+  arma::mat randomOutput = CLI::GetParam<arma::mat>("output");
+
+  SetInputParam<vector<string>>("input", {"test_image.png", "test_image.png"});
+  SetInputParam("input_model",
+                CLI::GetParam<ImageInfo*>("output_model"));
+
+  mlpackMain();
+  arma::mat savedOutput = CLI::GetParam<arma::mat>("output");
+  CheckMatrices(randomOutput, savedOutput);
+}
+
+/**
+ * Check whether binding throws error if height, width or channel are not
+ * specified.
+ */
+BOOST_AUTO_TEST_CASE(IncompleteTest)
+{
+  SetInputParam<vector<string>>("input", {"test_image.png", "test_image.png"});
+  SetInputParam("height", 50);
+  SetInputParam("width", 50);
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+ * Check for invalid height values.
+ */
+BOOST_AUTO_TEST_CASE(InvalidInputTest)
+{
+  SetInputParam<vector<string>>("input", {"test_image.png", "test_image.png"});
+  SetInputParam("height", -50);
+  SetInputParam("width", 50);
+  SetInputParam("channel", 3);
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+ * Check for invalid width values.
+ */
+BOOST_AUTO_TEST_CASE(InvalidWidthTest)
```

Comment:
I think it might be useful to add a test for when, e.g., a height/width/channel setting is specified that doesn't match the given input data.  Like, e.g., what happens if I pass `save = true` and a matrix with 100 rows, but then set height/width/channel to 50 each?

Reply:
It behaves weirdly, probably has to do something with related to STB

Reply:
shouldn't it be in load_save_test, Because Image converted just call the internal Load

Reply:
Ah, sure, I can agree with that.  The only thing is, it doesn't look like there's any code in `data::Save()` or `data::SaveImage()` to make that check, so I suppose we would have to add it there too. :+1:

Reply:
This is now fixed, thank you!  I'll resolve the conversation.